### ISSUE TITLE
Investigate problem with thumbnail appearing after login

### ISSUE THREAD

**Issue by pbinkley** [ualbertalib/HydraNorth#1169]
When you log in with your ccid, you often (always?) are taken to thumbnail image. This problem has been reported by a user as well as being observed by us. It appears that the thumbnail is for a ccid-protected item that appeared in search results, and that a request to download it was therefore unsuccessful, leading to a login screen (which wasn't shown to the user since it was a response to a request for an image). When the user does log in, possibly after a lot of other requests, they are taken to the thumbnail. But we think we've seen thumbnails from other sessions appear, so the problem appears not to be session-based (this needs confirmation).

Desired behavior: when a user clicks on a link to a protected item and is blocked, then logs in, we want to them to go to the item they were trying to get to. If a protected thumbnail is included in a search results page, though, we do not want it to affect the page the user goes to if they log in later in the session.

**Comment by leahvanderjagt** [comment]
Just to add: for me the image seems to be the same (Toronto Stock Exchange
data) whenever this happens, no matter what item I am attempting to access
via CCID. Do others also see the same image? Or does the thumbnail change
for others?

On Fri, Jun 3, 2016 at 4:07 PM, Peter Binkley <notifications@github.com>
wrote:

> When you log in with your ccid, you often (always?) are taken to thumbnail
> image. This problem has been reported by a user as well as being observed
> by us. It appears that the thumbnail is for a ccid-protected item that
> appeared in search results, and that a request to download it was therefore
> unsuccessful, leading to a login screen (which wasn't shown to the user
> since it was a response to a request for an image). When the user does log
> in, possibly after a lot of other requests, they are taken to the
> thumbnail. But we think we've seen thumbnails from other sessions appear,
> so the problem appears not to be session-based (this needs confirmation).
>
> Desired behavior: when a user clicks on a link to a protected item and is
> blocked, then logs in, we want to them to go to the item they were trying
> to get to. If a protected thumbnail is included in a search results page,
> though, we do not want it to affect the page the user goes to if they log
> in later in the session.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ualbertalib/HydraNorth/issues/1169>, or mute the
> thread
> <https://github.com/notifications/unsubscribe/AFieXMxMQunKrqLi-HKr8rMWdQ28JXeXks5qIKWUgaJpZM4It8oF>
> .
>



-- 
___________________________________________________________




*Leah VanderjagtDigital Repository Services Librarian University of
Albertat. 780.492.3851 *
*leahv@ualberta.ca <leahv@ualberta.ca> *

**Comment by pbinkley** [comment]
Related to #1159

**Comment by leahvanderjagt** [comment]
OK, I just logged in as admin using the non-CCID login method and I got the
attached screenshot. I was not even following a link, just logging into
ERA. This is such a weird bug!

On Fri, Jun 3, 2016 at 4:37 PM, Peter Binkley <notifications@github.com>
wrote:

> Related to #1159 <https://github.com/ualbertalib/HydraNorth/issues/1159>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ualbertalib/HydraNorth/issues/1169#issuecomment-223711195>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFieXIFhKunoOuBcX6egFiIp6BhCdaVXks5qIKylgaJpZM4It8oF>
> .
>



-- 
___________________________________________________________




*Leah VanderjagtDigital Repository Services Librarian University of
Albertat. 780.492.3851 *
*leahv@ualberta.ca <leahv@ualberta.ca> *

**Comment by leahvanderjagt** [comment]
<img width="587" alt="screen shot 2016-06-05 at 1 52 27 pm" src="https://cloud.githubusercontent.com/assets/5807708/15828444/8fe2c89e-2bcc-11e6-9642-355380fe1047.png">

**Comment by pgwillia** [comment]
@weiweishi this is why users are being taken to the thumbnail after login:

**Comment by pgwillia** [comment]
@weiweishi because the thumbnail is redirecting to the login page, the thumbnail is the stored_location for current scope and is why users are being taken to the thumbnail after login: 

```
  def after_sign_in_path_for(resource)
    if current_user.admin?
      stored_location_for(resource) || root_path
    else
      stored_location_for(resource) || sufia.dashboard_index_path || root_path
    end
  end
```
[https://github.com/ualbertalib/HydraNorth/blob/c6bf1d8242241a4444501ec7c503d4c93b4567f5/app/controllers/application_controller.rb#L51-L57]

```
# /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.4/lib/devise/controllers/store_location.rb
# Returns and delete (if it's navigational format) the url stored in the session for
      # the given scope. Useful for giving redirect backs after sign up:
      #
      # Example:
      #
      #   redirect_to stored_location_for(:user) || root_path
      #
      def stored_location_for(resource_or_scope)
        session_key = stored_location_key_for(resource_or_scope)

        if is_navigational_format?
          session.delete(session_key)
        else
          session[session_key]
        end
      end
```

**Comment by weiweishi** [comment]
@pgwillia that was my first thought as well, but then even the login is not issued from the home page, it still results in redirecting to the thumbnail after the authentication. That's where it gets confused as the thumbnail image for the recent object should not be in scope then.

**Comment by pbinkley** [comment]
So I think the [preferred solution](https://github.com/ualbertalib/HydraNorth/issues/1159#issuecomment-222755574) to #1159 would handle it: if a protected item would return an unprotected generic thumbnail image to non-logged-in users, the protected thumbnail url would never become the ```stored_location``` to confuse later login actions. I don't know how much work this would be though.

**Comment by weiweishi** [comment]
Are you okay to handle the logic in the view here @pgwillia
https://github.com/ualbertalib/HydraNorth/blob/master/app/views/recent/_recent_document.html.erb#L6-L8, any object that is not public, use the default generic thumbnail image.

**Comment by pgwillia** [comment]
@weiweishi I might want to put that logic in a helper.

**Comment by pbinkley** [comment]
:+1:  - we want to be sure the new logic applies anywhere the thumbnail is linked

**Comment by pgwillia** [comment]
# You can configure blacklight to use this as the thumbnail
    # example:
    #   config.index.thumbnail_method = :sufia_thumbnail_tag
    def sufia_thumbnail_tag(document, options)
[https://github.com/projecthydra/sufia/blob/6.x-stable/app/helpers/sufia/sufia_helper_behavior.rb#L39-L60]

**Comment by pgwillia** [comment]
#1179 might solve this too

**Comment by pbinkley** [comment]
Remaining bug fix in #1194
